src/teamcity.reporter.ts

```diff
@@ -1,113 +1,122 @@
-      if (parentSuite !== this.lastRunningSuite) {
-        if (typeof this.lastRunningSuite !== 'undefined') {
-          this.logSuiteResults(this.lastRunningSuite);
+      if (parentSuite !== this.#lastRunningSuite) {
+        if (this.#lastRunningSuite !== undefined) {
+          this.#logSuiteResults(this.#lastRunningSuite);
         }
-        this.lastRunningSuite = parentSuite;
+        this.#lastRunningSuite = parentSuite;
       }
     }
   }
 
   onEnd(result: FullResult) {
-    // @TODO try this
-    // https://www.jetbrains.com/help/teamcity/2021.2/service-messages.html#Importing+XML+Reports
-    // console.log(`##teamcity[importData type='junit' path='test-results.xml']`)
-
-    switch (this.report) {
-      case 'onTestBegin':
-        if (typeof this.lastRunningSuite !== 'undefined') {
-          this.logSuiteResults(this.lastRunningSuite);
+    switch (this.#mode) {
+      case ReporterMode.Test:
+        if (this.#lastRunningSuite !== undefined) {
+          this.#logSuiteResults(this.#lastRunningSuite);
         }
         break;
-      case 'onEnd':
+      case ReporterMode.Suite:
+      default:
+        // @TODO try this
+        // https://www.jetbrains.com/help/teamcity/2021.2/service-messages.html#Importing+XML+Reports
+        // console.log(`##teamcity[importData type='junit' path='test-results.xml']`);
         this.rootSuite.suites
-          .flatMap((projectSuite) => projectSuite.suites)
-          .forEach((suite) => this.logSuiteResults(suite));
+          .flatMap(fileSuite => fileSuite.suites)
+          .forEach(storySuite => this.#logSuiteResults(storySuite));
         break;
     }
-
     console.info(`Finished the run: ${result.status}`);
   }
 
-  private logSuiteResults(suite: Suite): void {
+  logToTC(action: ActionType, parts: string[]) {
+    const textParts = [
+      `##teamcity[${action}`,
+      ...parts.filter(part => !!part),
+      `flowId='${this.flowId}']`
+    ];
+    console.log(textParts.join(' '));
+  }
+
+  #logSuiteResults(suite: Suite): void {
     this.logToTC(`testSuiteStarted`, [
       `name='${TeamcityReporter.escape(suite.title)}'`
     ]);
-    suite.tests.forEach(t => this.logTestResults(t));
-    suite.suites.forEach(s => this.logSuiteResults(s));
+
+    suite.tests.forEach((testCase: TestCase) => this.#logTestResults(testCase));
+    suite.suites.forEach((suite: Suite) => this.#logSuiteResults(suite));
+
     this.logToTC(`testSuiteFinished`, [
       `name='${TeamcityReporter.escape(suite.title)}'`
     ]);
   }
 
-  private logTestResults(test: TestCase) {
-    const name = TeamcityReporter.escape(test.title);
-    const tzoffset = (new Date()).getTimezoneOffset() * 60000; //offset in milliseconds
-    const result = test.results.pop();
-    if (result === undefined) {
-      throw new Error(`Result should not be empty`);
+  #logTestResults(test: TestCase) {
+    const title = TeamcityReporter.escape(test.title);
+    if (process.env['PW_REPORT_All_RESULTS'] !== undefined) {
```

Comment:
@artemrudenko could this be moved to the constructor like I did with `TEAMCITY_ARTIFACTS_PW_RESULT` before?